docs/context-parameters.md

```diff
@@ -0,0 +1,41 @@
+Context Parameters
+==================
+
+KotlinPoet supports Kotlin's context parameters feature, which allows you to add context parameters to functions and properties. Context parameters are added using the `contextParameter()` method:
+
+```kotlin
+val greet = FunSpec.builder("greet")
+  .contextParameter("user", String::class)
+  .addStatement("println(\"Hello, ${'$'}user!\")")
+  .build()
+```
+
+The code above generates:
+
+```kotlin
+context(user: kotlin.String)
+public fun greet() {
+  println("Hello, $user!")
+}
+```
+
+You can add multiple context parameters:
+
+```kotlin
+val processData = FunSpec.builder("processData")
+  .contextParameter("logger", ClassName("com.example", "Logger"))
+  .contextParameter("config", ClassName("com.example", "Config"))
+  .addStatement("logger.info(\"Processing with config: ${'$'}config\")")
+  .build()
```

Comment:
Future improvement suggestion: it'd be cool to get `%N` to work with `ContextParameter`, so that we could do this:

```kotlin
val logger = ContextParameter("logger", Logger::class)
val config = ContextParameter("config", Config::class)
val processData = FunSpec.builder("processData")
  .contextParameter(logger)
  .contextParameter(config)
  .addStatement("%N.info(\"Processing with config: ${'$'}%N\")", logger, config)
  .build()
```

Let's land this separately though, this PR is starting to grow too large!

Reply:
Created ticket for this: https://github.com/square/kotlinpoet/issues/2114